autocertkit/utils.py

```diff
@@ -1914,54 +1908,65 @@ def __call__(self, *args):
             signal.signal(signal.SIGALRM, old)
         return result
 
-def xml_to_dicts(xml, tag):
-    doc = minidom.parseString(xml.strip())
-    try:
-        result = []
-        elems = doc.getElementsByTagName(tag)
-        for el in elems:
-            # Note that we have to convert this dictionary to non-unicode
-            # strings, because we're being casual elsewhere.  That's why we're
-            # not just returning dict(el.attributes.items()).
-            item = {}
-            for k, v in el.attributes.items():
-                item[str(k)] = str(v)
-            result.append(item)
-    finally:
-        doc.unlink()
-    log.debug(result)
-    return result
+def json_loads(json_data):
+    def process_dict_keys(d):
+        new_d = {}
+        for key in d.iterkeys():
+            new_key = str(key.replace(" ","_"))
+            new_d[new_key] = d[key]
+        return new_d
+
+    def process_values(item):
+        if isinstance(item, unicode):
+            item = str(item)
+        elif isinstance(item, list):
+            for elem in item:
+                elem = process_values(elem)
+        elif isinstance(item, dict):
+            item = process_dict_keys(item)
+            for key in item.iterkeys():
+                item[key] = str(item[key])
+        return item
+
+    data = json.loads(json_data, object_hook=process_values)
+    return [data] if isinstance(data, dict) else data
 
 def call_ack_plugin(session, method, args={}, host=None):
     if not host:
         host = get_pool_master(session)
     log.debug("About to call plugin '%s' on host '%s' with args '%s'" % \
                 (method, host, args))
-                
-    return session.xenapi.host.call_plugin(host,
+
+    res = session.xenapi.host.call_plugin(host,
                                            'autocertkit',
                                            method,
                                            args)
+    try:
+        res = json_loads(res)
+    except ValueError:
```

Comment:
Json.loads(res) throws ValueError when res is actually a string and not json. for example:
```python
>>> a= call_ack_plugin(session, 'get_kernel_version')
About to call plugin 'get_kernel_version' on host 'OpaqueRef:a9d94ee7-3999-cd74-6674-3788a6a78a72' with args '{}'
>>> a
'3.10.0+10'
>>> json.loads(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 368, in decode
    raise ValueError(errmsg("Extra data", s, end, len(s)))
ValueError: Extra data: line 1 column 5 - line 1 column 10 (char 4 - 9)
```